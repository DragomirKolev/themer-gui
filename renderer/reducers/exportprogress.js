import {
  EXPORT_CANCELLED,
  EXPORT_PROGRESS,
  EXPORT_ERROR,
  EXPORT_COMPLETE,
  EXPORT_RESET,
} from '../actions';
import {
  NOT_STARTED,
  IN_PROGRESS,
  COMPLETE_ERROR,
  COMPLETE_SUCCESS,
} from '../helpers/exportProgressStates';

const defaultExportProgress = {
  status: '',
  state: NOT_STARTED,
  exportedPath: null,
};

export const exportProgressReducer = (state = defaultExportProgress, action) => {
  switch (action.type) {
    case EXPORT_CANCELLED:
      return defaultExportProgress;
    case EXPORT_PROGRESS:
      return {
        ...defaultExportProgress,
        status: action.status,
        state: IN_PROGRESS,
      };
    case EXPORT_ERROR:
      return {
        ...defaultExportProgress,
        status: action.message,
        state: COMPLETE_ERROR,
      };
    case EXPORT_COMPLETE:
      return {
        status: `Your themes have been exported to ${action.exportedPath}`,
        state: COMPLETE_SUCCESS,
        exportedPath: action.exportedPath,
      };
    case EXPORT_RESET:
      return defaultExportProgress;
    default:
      return state;
  }
};
