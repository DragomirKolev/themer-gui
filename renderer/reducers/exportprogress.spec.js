import { exportProgressReducer } from './exportprogress';
import {
  exportCancelled,
  exportProgress,
  exportError,
  exportComplete,
  exportReset,
} from '../actions';
import {
  NOT_STARTED,
  IN_PROGRESS,
  COMPLETE_ERROR,
  COMPLETE_SUCCESS,
} from '../helpers/exportProgressStates';

describe('exportProgressReducer', () => {
  it('should initialize to a state of not started', () => {
    expect(exportProgressReducer(undefined, {})).toMatchSnapshot();
  });
  it('should reset when the export is cancelled', () => {
    expect(exportProgressReducer(undefined, exportCancelled())).toMatchSnapshot();
  });
  it('should reflect current status when there is progress', () => {
    expect(exportProgressReducer(undefined, exportProgress('working...'))).toMatchSnapshot();
  });
  it('should support an error state', () => {
    expect(exportProgressReducer(undefined, exportError('something went wrong'))).toMatchSnapshot();
  });
  it('should support complete state with path', () => {
    expect(exportProgressReducer(undefined, exportComplete('/dev/null'))).toMatchSnapshot();
  });
  it('should properly reset when needed', () => {
    const beginningState = {
      status: 'working...',
      state: IN_PROGRESS,
      exportedPath: null,
    };
    const expected = exportProgressReducer(undefined, {});
    expect(exportProgressReducer(beginningState, exportReset())).toEqual(expected);
  });
});
