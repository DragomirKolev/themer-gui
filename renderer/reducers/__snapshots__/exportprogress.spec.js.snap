// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exportProgressReducer should initialize to a state of not started 1`] = `
Object {
  "exportedPath": null,
  "state": "NOT_STARTED",
  "status": "",
}
`;

exports[`exportProgressReducer should reflect current status when there is progress 1`] = `
Object {
  "exportedPath": null,
  "state": "IN_PROGRESS",
  "status": "working...",
}
`;

exports[`exportProgressReducer should reset when the export is cancelled 1`] = `
Object {
  "exportedPath": null,
  "state": "NOT_STARTED",
  "status": "",
}
`;

exports[`exportProgressReducer should support an error state 1`] = `
Object {
  "exportedPath": null,
  "state": "COMPLETE_ERROR",
  "status": "something went wrong",
}
`;

exports[`exportProgressReducer should support complete state with path 1`] = `
Object {
  "exportedPath": "/dev/null",
  "state": "COMPLETE_SUCCESS",
  "status": "Your themes have been exported to /dev/null",
}
`;
